src/platforms/hosted/dap.c

```diff
@@ -583,40 +594,41 @@ void dap_ap_write(adiv5_access_port_s *ap, uint16_t addr, uint32_t value)
 
 void dap_read_single(adiv5_access_port_s *ap, void *dest, uint32_t src, align_e align)
 {
-	uint8_t buf[64];
-	const uint32_t length = mem_access_setup(ap, buf, src, align);
-	buf[length] = SWD_AP_DRW | DAP_TRANSFER_RnW;
-	buf[length + 1] = SWD_DP_R_RDBUFF | DAP_TRANSFER_RnW;
-	buf[2] += 2; /* Increase the number of transactions by 2 */
-	uint32_t tmp = wait_word(buf, 64, length + 2U, &ap->dp->fault);
-	dest = extract(dest, src, tmp, align);
+	dap_transfer_request_s requests[5];
+	mem_access_setup(ap, requests, src, align);
+	requests[3].request = SWD_AP_DRW | DAP_TRANSFER_RnW;
+	requests[4].request = SWD_DP_R_RDBUFF | DAP_TRANSFER_RnW;
+	uint32_t result[2];
+	if (!perform_dap_transfer(ap->dp, requests, 5U, result, 2U)) {
+		DEBUG_WARN("dap_read_single failed\n");
+		memset(dest, 0, sizeof(uint32_t));
+		return;
+	}
+	/* Pull out the data from the *second* response (value of RDBUFF) */
```

Comment:
This is incorrect. CMSIS-DAP handles the AP read latency on the probe side and does an implicit read of RDBUFF, so the host should never need to read it explicitly. The actual read value ends up in the first response slot, and the second contains undefined garbage (all zeroes in my testing). The proper solution here is to simply remove the fifth transfer request and return the first response.

Reply:
Many thanks for the extra information, this should now be fixed